File: app/src/main/java/com/catsmoker/app/PlayTimeTrackerService.java
package com.catsmoker.app;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.app.usage.UsageStats;
import android.app.usage.UsageStatsManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import java.util.List;
import java.util.SortedMap;
import java.util.TreeMap;

public class PlayTimeTrackerService extends Service {

    public static final String EXTRA_PACKAGE_NAME = "extra_package_name";
    private static final String TAG = "PlayTimeTrackerService";
    private static final String NOTIFICATION_CHANNEL_ID = "PlayTimeTrackerChannel";
    private static final int NOTIFICATION_ID = 1;

    private Handler handler;
    private Runnable runnable;
    private String trackedPackageName;
    private long sessionPlayTime = 0;

    @Override
    public void onCreate() {
        super.onCreate();
        handler = new Handler(Looper.getMainLooper());
        createNotificationChannel();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null && intent.hasExtra(EXTRA_PACKAGE_NAME)) {
            trackedPackageName = intent.getStringExtra(EXTRA_PACKAGE_NAME);
            sessionPlayTime = 0;

            Notification notification = new NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
                    .setContentTitle("Tracking Play Time")
                    .setContentText("Tracking play time for " + trackedPackageName)
                    .setSmallIcon(R.drawable.ic_check_circle)
                    .build();

            startForeground(NOTIFICATION_ID, notification);

            startTracking();
        }
        return START_STICKY;
    }

    private void startTracking() {
        runnable = new Runnable() {
            @Override
            public void run() {
                if (isPackageInForeground()) {
                    sessionPlayTime++;
                    Log.d(TAG, "Play time for " + trackedPackageName + ": " + sessionPlayTime + "s");
                } else {
                    stopTrackingAndSelf();
                }
                handler.postDelayed(this, 1000);
            }
        };
        handler.post(runnable);
    }

    private void stopTrackingAndSelf() {
        handler.removeCallbacks(runnable);
        savePlayTime();
        stopForeground(true);
        stopSelf();
    }

    private boolean isPackageInForeground() {
        UsageStatsManager usageStatsManager = (UsageStatsManager) getSystemService(Context.USAGE_STATS_SERVICE);
        long time = System.currentTimeMillis();
        List<UsageStats> stats = usageStatsManager.queryUsageStats(UsageStatsManager.INTERVAL_DAILY, time - 1000 * 10, time);
        if (stats != null) {
            SortedMap<Long, UsageStats> mySortedMap = new TreeMap<>();
            for (UsageStats usageStats : stats) {
                mySortedMap.put(usageStats.getLastTimeUsed(), usageStats);
            }
            if (!mySortedMap.isEmpty()) {
                String currentApp = mySortedMap.get(mySortedMap.lastKey()).getPackageName();
                return currentApp.equals(trackedPackageName);
            }
        }
        return false;
    }

    private void savePlayTime() {
        if (trackedPackageName != null) {
            SharedPreferences prefs = getSharedPreferences("PlayTime", MODE_PRIVATE);
            long totalPlayTime = prefs.getLong(trackedPackageName, 0);
            totalPlayTime += sessionPlayTime;
            prefs.edit().putLong(trackedPackageName, totalPlayTime).apply();
        }
    }

    @Override
    public void onDestroy() {
        stopTrackingAndSelf();
        super.onDestroy();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel serviceChannel = new NotificationChannel(
                    NOTIFICATION_CHANNEL_ID,
                    "Play Time Tracker",
                    NotificationManager.IMPORTANCE_DEFAULT
            );
            NotificationManager manager = getSystemService(NotificationManager.class);
            manager.createNotificationChannel(serviceChannel);
        }
    }
}

File: app/src/main/java/com/catsmoker/app/gametools/PlayTimeTrackerService.kt
package com.catsmoker.app.gametools

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.app.usage.UsageStats
import android.app.usage.UsageStatsManager
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.content.edit
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

class PlayTimeTrackerService : Service() {

    private val serviceScope = CoroutineScope(Dispatchers.Main)
    private var trackingJob: Job? = null

    private var trackedPackageName: String? = null
    private var sessionPlayTime: Long = 0
    private var vpnEnabledForSession = false

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent == null || !intent.hasExtra(EXTRA_PACKAGE_NAME)) {
            stopSelf()
            return START_NOT_STICKY
        }

        trackedPackageName = intent.getStringExtra(EXTRA_PACKAGE_NAME)
        vpnEnabledForSession = intent.getBooleanExtra(EXTRA_VPN_ENABLED, false)
        sessionPlayTime = 0

        val notification = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
            .setContentTitle("Game Booster Running")
            .setContentText("Tracking play time for ${getAppName(trackedPackageName ?: "Unknown")}")
            .setSmallIcon(android.R.drawable.ic_media_play)
            .setOngoing(true)
            .setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)
            .build()

        if (Build.VERSION.SDK_INT >= 34) {
            startForeground(NOTIFICATION_ID, notification, ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE)
        } else if (Build.VERSION.SDK_INT >= 29) {
            startForeground(NOTIFICATION_ID, notification, 0)
        } else {
            startForeground(NOTIFICATION_ID, notification)
        }

        startTracking()
        return START_STICKY
    }

    private fun startTracking() {
        trackingJob?.cancel()
        trackingJob = serviceScope.launch(Dispatchers.IO) {
            while (isActive) {
                if (isPackageInForeground) {
                    sessionPlayTime += (CHECK_INTERVAL_MS / 1000)
                    Log.d(TAG, "Play time for $trackedPackageName: ${sessionPlayTime}s")
                    delay(CHECK_INTERVAL_MS)
                } else {
                    Log.i(TAG, "Game no longer in foreground. Stopping tracking.")
                    stopTrackingAndSelf()
                    break
                }
            }
        }
    }

    private fun stopTrackingAndSelf() {
        trackingJob?.cancel()
        savePlayTime()

        if (vpnEnabledForSession) {
            val vpnIntent = Intent(this, GameVpnService::class.java).apply {
                action = GameVpnService.ACTION_DISCONNECT
            }
            startService(vpnIntent)
        }

        stopForeground(STOP_FOREGROUND_REMOVE)
        stopSelf()
    }

    private val isPackageInForeground: Boolean
        get() {
            val targetPackage = trackedPackageName ?: return false

            val usageStatsManager = getSystemService(USAGE_STATS_SERVICE) as? UsageStatsManager ?: return false
            val time = System.currentTimeMillis()

            val stats = usageStatsManager.queryUsageStats(
                UsageStatsManager.INTERVAL_DAILY,
                time - (CHECK_INTERVAL_MS * 2),
                time
            )

            val latest = stats?.maxByOrNull { it.lastTimeUsed } ?: return false
            return targetPackage == latest.packageName
        }

    private fun savePlayTime() {
        val packageName = trackedPackageName ?: return
        if (sessionPlayTime <= 0) return
        val prefs = getSharedPreferences(PREFS_PLAY_TIME, MODE_PRIVATE)
        val totalPlayTime = prefs.getLong(packageName, 0) + sessionPlayTime
        prefs.edit { putLong(packageName, totalPlayTime) }
        Log.i(TAG, "Saved ${sessionPlayTime}s for $packageName. Total: ${totalPlayTime}s")
    }

    private fun getAppName(packageName: String): String {
        return try {
            packageManager.getApplicationLabel(
                packageManager.getApplicationInfo(packageName, 0)
            ).toString()
        } catch (_: Exception) {
            packageName
        }
    }

    override fun onDestroy() {
        // Ensure time is saved if killed
        if (sessionPlayTime > 0) {
            savePlayTime()
        }
        serviceScope.cancel()
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null

    private fun createNotificationChannel() {
        val serviceChannel = NotificationChannel(
            NOTIFICATION_CHANNEL_ID,
            "Play Time Tracker",
            NotificationManager.IMPORTANCE_LOW
        ).apply {
            description = "Shows when a game is being tracked"
        }
        getSystemService(NotificationManager::class.java)?.createNotificationChannel(serviceChannel)
    }

    companion object {
        const val EXTRA_PACKAGE_NAME = "extra_package_name"
        const val EXTRA_VPN_ENABLED = "extra_vpn_enabled"

        private const val TAG = "PlayTimeTrackerService"
        private const val PREFS_PLAY_TIME = "PlayTime"
        private const val NOTIFICATION_CHANNEL_ID = "PlayTimeTrackerChannel"
        private const val NOTIFICATION_ID = 1
        private const val CHECK_INTERVAL_MS = 2000L
    }
}
